bridge/ffi/src/c_api.rs

```diff
@@ -0,0 +1,203 @@
+	let payloads = payloads_from_ffi(&payloads);
+	graph_state.import_users_data(&payloads).is_ok()
+}
+
+// Graph export updates
+#[no_mangle]
+pub unsafe extern "C" fn graph_export_updates(graph_state: *mut GraphState) -> GraphUpdates {
+	let graph_state = &mut *graph_state;
+	let updates = graph_state.export_updates().unwrap_or_default();
+	let ffi_updates = updates_to_ffi(updates);
+	let updates_len = ffi_updates.len();
+	let updates_ptr = ManuallyDrop::new(ffi_updates).as_mut_ptr();
+	GraphUpdates { updates: updates_ptr, updates_len }
+}
+
+// Graph apply actions
+#[no_mangle]
+pub unsafe extern "C" fn graph_apply_actions(
+	graph_state: *mut GraphState,
+	actions: *const Action,
+	actions_len: usize,
+) -> bool {
+	let graph_state = &mut *graph_state;
+	let actions = std::slice::from_raw_parts(actions, actions_len);
+	graph_state.apply_actions(&actions).is_ok()
+}
+
+// Graph get connections for user
+#[no_mangle]
+pub unsafe extern "C" fn graph_get_connections_for_user(
+	graph_state: *mut GraphState,
+	user_id: *const DsnpUserId,
+	schema_id: *const SchemaId,
+	include_pending: bool,
+) -> GraphConnections {
+	let graph_state = &mut *graph_state;
+	let user_id = &*user_id;
+	let schema_id = &*schema_id;
+	let connections = graph_state
+		.get_connections_for_user_graph(user_id, schema_id, include_pending)
+		.unwrap_or_default();
+	let connections_len = connections.len();
+	let connections_ptr = ManuallyDrop::new(connections).as_mut_ptr();
+	GraphConnections { connections: connections_ptr, connections_len }
+}
+
+// Get connections without keys
+#[no_mangle]
+pub unsafe extern "C" fn graph_get_connections_without_keys(
+	graph_state: *mut GraphState,
+) -> GraphConnectionsWithoutKeys {
+	let graph_state = &mut *graph_state;
+	let connections = graph_state.get_connections_without_keys().unwrap_or_default();
+	let connections_len = connections.len();
+	let connections_ptr = ManuallyDrop::new(connections).as_mut_ptr();
+	GraphConnectionsWithoutKeys { connections: connections_ptr, connections_len }
+}
+
+// Get one sided private friendship connections
+#[no_mangle]
+pub unsafe extern "C" fn graph_get_one_sided_private_friendship_connections(
+	graph_state: *mut GraphState,
+	user_id: *const DsnpUserId,
+) -> GraphConnections {
+	let graph_state = &mut *graph_state;
+	let user_id = &*user_id;
+	let connections = graph_state
+		.get_one_sided_private_friendship_connections(user_id)
+		.unwrap_or_default();
+	let connections_len = connections.len();
+	let connections_ptr = ManuallyDrop::new(connections).as_mut_ptr();
+	GraphConnections { connections: connections_ptr, connections_len }
+}
+
+// Free GraphState
+#[no_mangle]
+pub unsafe extern "C" fn free_graph_state(graph_state: *mut GraphState) {
+	let _ = Box::from_raw(graph_state);
+}
+
+// Free GraphStates
+#[no_mangle]
+pub unsafe extern "C" fn free_graph_states() {
+	for graph_state in GRAPH_STATES.iter() {
+		let _ = Box::from_raw(*graph_state);
+	}
+}
```

Comment:
If you free an individual graph, but don't remove it from the GRAPH_STATES vector, you're in danger of a double-free if you later call `free_graph_states()`

Reply:
We also might want some kind of thread-safe vector for GRAPH_STATES; mutable static is not thread-safe.

Reply:
...and we might want to try and prevent a client double-free from calling `free_graph_state()` on the same pointer twice:
1. Check that pointer is in GRAPH_STATES vector before freeing
2. Take pointer from GRAPH_STATES vector
3. Free

Reply:
...or we could change GRAPH_STATES to a HashMap with integer keys, and instead of using pointers, the client just gets a graph ID to specify which graph...

Reply:
:point_up: would still need to be thread-safe